Add flux & dirichlet boundary conditions for mass/momentum/energy in NS

GiudGiud opened the pull request:
closes #19543

Needs docs & tests. 
This PR should also have the scalars while we're at it.

GiudGiud committed:
Add flux & dirichlet boundary conditions for SAM-Pgh coupling, closes #19543

GiudGiud commented:
where's the mark as draft button again

moosebuild commented:
Job [Precheck](https://civet.inl.gov/job/910252/) on 8f4a72f wanted to post the following:

Your code requires style changes.  

A patch was auto generated and copied [here](https://mooseframework.inl.gov/docs/PRs/19568/style.patch)  
You can directly apply the patch by running, in the top level of your repository:

`curl -s https://mooseframework.inl.gov/docs/PRs/19568/style.patch | git apply -v`  

Alternatively, with your repository up to date and in the top level of your repository:

`git clang-format 19a82af2614e795c0705daab8b93202b418d184a`

moosebuild commented:
Job [Documentation](https://civet.inl.gov/job/910310/) on 541873a wanted to post the following:

View the site [here](https://mooseframework.inl.gov/docs/PRs/19568/site)

This comment will be updated on new commits.

moosebuild commented:
Job [Coverage](https://civet.inl.gov/job/910320/) on 541873a wanted to post the following:


## Framework coverage

<table>
  <thead>
    <tr>
      <th></th>
      <th>d3a9b3</th>
      <th colspan=4>#19568 541873</th>
    </tr>
    <tr>
      <th></th>
      <th>Total</th>
      <th>Total</th>
      <th>+/-</th>
      <th>New</th>
    </tr>
  </thead>
  <tbody>
    <tr>
      <td><b>Rate</b></td>
      <td>82.10%</td>
      <td>82.10%</td>
      <td>-0.00%</td>
      <td>-</td>
    </tr>
    <tr>
      <td><b>Hits</b></td>
      <td>72724</td>
      <td>72723</td>
      <td>-1</td>
      <td>0</td>
    </tr>
    <tr>
      <td><b>Misses</b></td>
      <td>15851</td>
      <td>15852</td>
      <td>+1</td>
      <td>0</td>
    </tr>
  </tbody>
</table>

[Diff coverage report](https://mooseframework.inl.gov/docs/PRs/19568/coverage_diff/framework.html)

[Full coverage report](https://mooseframework.inl.gov/docs/PRs/19568/coverage/framework)


## Modules coverage

### Navier stokes

<table>
  <thead>
    <tr>
      <th></th>
      <th>d3a9b3</th>
      <th colspan=4>#19568 541873</th>
    </tr>
    <tr>
      <th></th>
      <th>Total</th>
      <th>Total</th>
      <th>+/-</th>
      <th>New</th>
    </tr>
  </thead>
  <tbody>
    <tr>
      <td><b>Rate</b></td>
      <td>83.17%</td>
      <td>81.99%</td>
      <td>-1.17%</td>
      <td>26.04%</td>
    </tr>
    <tr>
      <td><b>Hits</b></td>
      <td>7609</td>
      <td>7659</td>
      <td>+50</td>
      <td>50</td>
    </tr>
    <tr>
      <td><b>Misses</b></td>
      <td>1540</td>
      <td>1682</td>
      <td>+142</td>
      <td>142</td>
    </tr>
  </tbody>
</table>

[Diff coverage report](https://mooseframework.inl.gov/docs/PRs/19568/coverage_diff/navier_stokes.html)

[Full coverage report](https://mooseframework.inl.gov/docs/PRs/19568/coverage/navier_stokes)

## Full coverage reports

<details>
<summary>Reports</summary>

- [`framework`](https://mooseframework.inl.gov/docs/PRs/19568/coverage/framework)
- [`chemical_reactions`](https://mooseframework.inl.gov/docs/PRs/19568/coverage/chemical_reactions)
- [`combined`](https://mooseframework.inl.gov/docs/PRs/19568/coverage/combined)
- [`contact`](https://mooseframework.inl.gov/docs/PRs/19568/coverage/contact)
- [`external_petsc_solver`](https://mooseframework.inl.gov/docs/PRs/19568/coverage/external_petsc_solver)
- [`fluid_properties`](https://mooseframework.inl.gov/docs/PRs/19568/coverage/fluid_properties)
- [`fsi`](https://mooseframework.inl.gov/docs/PRs/19568/coverage/fsi)
- [`functional_expansion_tools`](https://mooseframework.inl.gov/docs/PRs/19568/coverage/functional_expansion_tools)
- [`geochemistry`](https://mooseframework.inl.gov/docs/PRs/19568/coverage/geochemistry)
- [`heat_conduction`](https://mooseframework.inl.gov/docs/PRs/19568/coverage/heat_conduction)
- [`level_set`](https://mooseframework.inl.gov/docs/PRs/19568/coverage/level_set)
- [`misc`](https://mooseframework.inl.gov/docs/PRs/19568/coverage/misc)
- [`navier_stokes`](https://mooseframework.inl.gov/docs/PRs/19568/coverage/navier_stokes)
- [`peridynamics`](https://mooseframework.inl.gov/docs/PRs/19568/coverage/peridynamics)
- [`phase_field`](https://mooseframework.inl.gov/docs/PRs/19568/coverage/phase_field)
- [`porous_flow`](https://mooseframework.inl.gov/docs/PRs/19568/coverage/porous_flow)
- [`ray_tracing`](https://mooseframework.inl.gov/docs/PRs/19568/coverage/ray_tracing)
- [`rdg`](https://mooseframework.inl.gov/docs/PRs/19568/coverage/rdg)
- [`reactor`](https://mooseframework.inl.gov/docs/PRs/19568/coverage/reactor)
- [`richards`](https://mooseframework.inl.gov/docs/PRs/19568/coverage/richards)
- [`stochastic_tools`](https://mooseframework.inl.gov/docs/PRs/19568/coverage/stochastic_tools)
- [`tensor_mechanics`](https://mooseframework.inl.gov/docs/PRs/19568/coverage/tensor_mechanics)
- [`xfem`](https://mooseframework.inl.gov/docs/PRs/19568/coverage/xfem)

</details>

## Warnings

- `navier_stokes` new line coverage rate 26.04% is less than the suggested 90.0%



This comment will be updated on new commits.

GiudGiud committed:
Add WCNSFV scalar flux BCs for scalars, such as the dnps, being advected through the system

GiudGiud committed:
Add symmetry BC for scalar advection

GiudGiud committed:
Add documentation for the WCNSFV flux boundary conditions

GiudGiud committed:
Add documentation for dirichlet BCs for fluid flow

GiudGiud committed:
Add tests for finite volume flux and dirichlet BCs
Clang format

moosebuild commented:
Job [Precheck](https://civet.inl.gov/job/915444/) on fe52d74 wanted to post the following:

Your code requires style changes.  

A patch was auto generated and copied [here](https://mooseframework.inl.gov/docs/PRs/19568/style.patch)  
You can directly apply the patch by running, in the top level of your repository:

`curl -s https://mooseframework.inl.gov/docs/PRs/19568/style.patch | git apply -v`  

Alternatively, with your repository up to date and in the top level of your repository:

`git clang-format a7794ad7aaf012a2e54bdaf98ea3ebc722a794f3`

GiudGiud committed:
Fixup tests for new BCs
Fixup exceptions for insufficient information in BCs
Add check for 0 surface area
refs #19543

Lighten test exodus files

test fix

GiudGiud committed:
Add class descriptions
fix test failure: looser tolerance on one case
refs #19543

GiudGiud commented:
With an isothermal region at the beginning we could re-use the same exodus 5 times I think.
We could also cut down on the exodus file size by looking only at the final step?

lemme know what you think

lindsayad commented:
I can't hear anything until the interpolator PR is merged 😆 

But yea I am game for reusing gold files and limiting the size of said files

GiudGiud commented:
the only issue is that by erasing the differences between those ways of doing the BC, we are purposefully not testing the case where they perform differently. 
Questions why we have a test suite, to get coverage on those lines, to make sure no one completely breaks the code, or to test subtle differences.

Anyway, yeah I know. Ill get to it now. This is needed for SAM-pgh so I wanted to get it in a workable shape

lindsayad commented:
I haven't looked at this so I basically have zero understanding of what's going on at this point. Based on what you just said I would amend my previous comment and say I think we should have 5 different tests with 5 different gold files. Or at least it seems like there should be because I presume we should have 5 different requirements

lindsayad commented on modules/navier_stokes/include/fvbcs/WCNSFVEnergyFluxBC.h:
```diff
@@ -0,0 +1,50 @@
+  const Real _scaling_factor;
+
+  /// Postprocessor with the inlet temperature
+  const PostprocessorValue * _temperature_pp;
+
+  /// Postprocessor with the inlet energy flow rate
+  const PostprocessorValue * _energy_pp;
+
+  /// Postprocessor with the inlet velocity
+  const PostprocessorValue * _velocity_pp;
+
+  /// Postprocessor with the inlet mass flow rate
+  const PostprocessorValue * _mdot_pp;
+
+  /// Postprocessor with the inlet area
+  const PostprocessorValue * _area_pp;
+
+  /// Fluid density functor
+  const Moose::Functor<ADReal> * _rho;
+
+  /// Fluid specific heat capacity functor
+  const Moose::Functor<ADReal> * _cp;
```
```suggestion
  const PostprocessorValue * const _temperature_pp;

  /// Postprocessor with the inlet energy flow rate
  const PostprocessorValue * const _energy_pp;

  /// Postprocessor with the inlet velocity
  const PostprocessorValue * const _velocity_pp;

  /// Postprocessor with the inlet mass flow rate
  const PostprocessorValue * const _mdot_pp;

  /// Postprocessor with the inlet area
  const PostprocessorValue * const _area_pp;

  /// Fluid density functor
  const Moose::Functor<ADReal> * const _rho;

  /// Fluid specific heat capacity functor
  const Moose::Functor<ADReal> * const _cp;
```

lindsayad commented on modules/navier_stokes/include/fvbcs/WCNSFVInletTemperatureBC.h:
```diff
@@ -0,0 +1,52 @@
+  /// Scaling factor
+  const Real _scaling_factor;
+
+  /// Postprocessor with the inlet temperature
+  const PostprocessorValue * _temperature_pp;
+
+  /// Postprocessor with the inlet energy flow rate
+  const PostprocessorValue * _energy_pp;
+
+  /// Postprocessor with the inlet velocity
+  const PostprocessorValue * _velocity_pp;
+
+  /// Postprocessor with the inlet mass flow rate
+  const PostprocessorValue * _mdot_pp;
+
+  /// Postprocessor with the inlet area
+  const PostprocessorValue * _area_pp;
+
+  /// Fluid density functor
+  const Moose::Functor<ADReal> * _rho;
+
+  /// Fluid specific heat capacity functor
+  const Moose::Functor<ADReal> * _cp;
```
all pointers set in in the initialization list should be marked `const`

lindsayad commented on modules/navier_stokes/include/fvbcs/WCNSFVInletVelocityBC.h:
```diff
@@ -0,0 +1,43 @@
+  /// Scaling factor
+  const Real _scaling_factor;
+
+  /// Postprocessor with the inlet velocity
+  const PostprocessorValue * _velocity_pp;
+
+  /// Postprocessor with the inlet mass flow rate
+  const PostprocessorValue * _mdot_pp;
+
+  /// Postprocessor with the inlet area
+  const PostprocessorValue * _area_pp;
+
+  /// Fluid density functor
+  const Moose::Functor<ADReal> * _rho;
```
all pointers set in in the initialization list should be marked `const`

lindsayad commented on modules/navier_stokes/include/fvbcs/WCNSFVMassFluxBC.h:
```diff
@@ -0,0 +1,41 @@
+  /// Scaling factor
+  const Real _scaling_factor;
+
+  /// Postprocessor with the inlet velocity
+  const PostprocessorValue * _velocity_pp;
+
+  /// Postprocessor with the inlet mass flow rate
+  const PostprocessorValue * _mdot_pp;
+
+  /// Postprocessor with the inlet area
+  const PostprocessorValue * _area_pp;
+
+  /// Fluid density functor
+  const Moose::Functor<ADReal> * _rho;
```
all pointers set in in the initialization list should be marked `const`

lindsayad commented on modules/navier_stokes/include/fvbcs/WCNSFVMomentumFluxBC.h:
```diff
@@ -0,0 +1,41 @@
+  /// Scaling factor
+  const Real _scaling_factor;
+
+  /// Postprocessor with the inlet velocity
+  const PostprocessorValue * _velocity_pp;
+
+  /// Postprocessor with the inlet mass flow rate
+  const PostprocessorValue * _mdot_pp;
+
+  /// Postprocessor with the inlet area
+  const PostprocessorValue * _area_pp;
+
+  /// Fluid density functor
+  const Moose::Functor<ADReal> * _rho;
```
all pointers set in in the initialization list should be marked `const`

lindsayad commented on modules/navier_stokes/include/fvbcs/WCNSFVScalarFluxBC.h:
```diff
@@ -0,0 +1,47 @@
+  /// Scaling factor
+  const Real _scaling_factor;
+
+  /// Postprocessor with the inlet scalar concentration
+  const PostprocessorValue * _scalar_value_pp;
+
+  /// Postprocessor with the inlet scalar flow rate
+  const PostprocessorValue * _scalar_flux_pp;
+
+  /// Postprocessor with the inlet velocity
+  const PostprocessorValue * _velocity_pp;
+
+  /// Postprocessor with the inlet mass flow rate
+  const PostprocessorValue * _mdot_pp;
+
+  /// Postprocessor with the inlet area
+  const PostprocessorValue * _area_pp;
+
+  /// Fluid density functor
+  const Moose::Functor<ADReal> * _rho;
```
all pointers set in in the initialization list should be marked `const`

lindsayad commented on modules/navier_stokes/src/fvbcs/INSFVSymmetryPressureBC.C:
```diff
@@ -11,12 +11,13 @@
   auto params = INSFVSymmetryBC::validParams();
-  params.addClassDescription("Though applied to the pressure, this object ensures that the "
+  params.addClassDescription("Though not applied to velocity, this object ensures that the "
                              "velocity perpendicular to a symmetry bounadry is zero by setting the "
                              "mass flow rate across the symmetry boundary to zero.");
```
The rest of this description is not accurate if the variable is not pressure

GiudGiud replied:
ok modified a bit.

lindsayad replied:
Love the newest description

lindsayad commented on modules/navier_stokes/src/fvbcs/WCNSFVInletTemperatureBC.C:
```diff
@@ -0,0 +1,134 @@
+    ADReal rho;
+    ADReal cp;
+
+    // TODO: Make this a utility to all FVBCs
+    const bool use_elem = fi.faceType(_var.name()) == FaceInfo::VarFaceNeighbors::ELEM;
+
+    if (use_elem)
+    {
+      rho = (*_rho)(std::make_tuple(
+          &fi, Moose::FV::LimiterType::CentralDifference, true, fi.elem().subdomain_id()));
+      cp = (*_cp)(std::make_tuple(
+          &fi, Moose::FV::LimiterType::CentralDifference, true, fi.elem().subdomain_id()));
+    }
+    else
+    {
+      rho = (*_rho)(std::make_tuple(
+          &fi, Moose::FV::LimiterType::CentralDifference, true, fi.neighborPtr()->subdomain_id()));
+      cp = (*_cp)(std::make_tuple(
+          &fi, Moose::FV::LimiterType::CentralDifference, true, fi.neighborPtr()->subdomain_id()));
+    }
+
+    return _scaling_factor * (*_energy_pp) / (*_area_pp * rho.value() * *_velocity_pp * cp.value());
```
This is what `singleSidedFaceArg` does. I would suggest moving that method and the `_face_info` member down to `FVBoundaryCondition`, set `_face_info = &fi` in this method, and then call `singleSidedFaceArg`. Similarly for all the other times this is done in this PR

GiudGiud replied:
true true. it s what I used for the flux BCs. 

Any reason the DirichletBCs are Real and not ADReal?

lindsayad replied:
In the beginning we only imagined fairly simple Dirichlet conditions: e.g. constants and direct functions of space and time. But now if you have more advanced conditions that could involve another variables nonlinear degrees of freedom, then we could change it to return `ADReal` for sure

GiudGiud replied:
so I remember why I did not do that now. computeValue is const, so I cant assign _face_info.

GiudGiud replied:
so either we
- keep it as I have
- assign _face_info earlier, like before computingValue (well I think i like that most)
- make all computeValue not const

GiudGiud replied:
Moving up only singleSidedFaceArgs to FVBoundaryCondition feels like all these routines should move together.
Duplicating it into FVDirichletBCBase isnt great either though

GiudGiud replied:
or we could just add a non-const version of boundaryValue. Seems to defeat the purpose of const-ness most of the time though.

lindsayad replied:
What are you including in "all these routines"? I'm a fan of moving `singleSidedFaceArg` up to `FVBoundaryCondition` and then adding a default `const FaceInfo * face_info = nullptr` argument that if `nullptr` automatically resolves to `_face_info`. This is what is done in `FVFluxKernel::faceArgSubdomains` (well at least on the interpolator branch but I think in upstream as well)

lindsayad replied:
So I guess that also would necessitate moving `_face_info` up to `FVBoundaryCondition` as well (which seems a natural move to me)

GiudGiud replied:
> so I remember why I did not do that now. boundaryValue is const, so I cant assign _face_info.

This is the problem really. Moving attributes is ok imo.

lindsayad replied:
You don't need `_face_info` assigned if you add a `face_info` parameter to `singleSidedFaceArg`

GiudGiud replied:
sounds good

GiudGiud committed:
Address Alex's review
- make pointers to postprocessors const
- move face_info up to FVBoundaryCondition, allowing singleSidedFaceArgs to move up too
- use the singleSidedFaceArgs in the inlet dirichlet BCs
Fix tests
- restrict ad indexing type for ewception testing
- use the right headers for non-unity build
Refs #19543

Apply suggestions from code review

Co-authored-by: Alex Lindsay <alexlindsay239@gmail.com>

lindsayad commented on framework/include/fvbcs/FVDirichletBCBase.h:
```diff
@@ -24,4 +24,16 @@ class FVDirichletBCBase : public FVBoundaryCondition
+   *        interpolation is required for the parameters of the functor
+   */
+  std::tuple<const FaceInfo *, Moose::FV::LimiterType, bool, SubdomainID> singleSidedFaceArg(
+      const FaceInfo * fi,
```
```suggestion
      const FaceInfo * fi = nullptr,
```

lindsayad commented on framework/src/fvbcs/FVDirichletBCBase.C:
```diff
@@ -22,3 +22,15 @@ FVDirichletBCBase::FVDirichletBCBase(const InputParameters & parameters)
+FVDirichletBCBase::singleSidedFaceArg(const FaceInfo * fi,
+                                      Moose::FV::LimiterType limiter_type) const
+{
+  const bool use_elem = fi->faceType(_var.name()) == FaceInfo::VarFaceNeighbors::ELEM;
```
```suggestion
  if (!fi)
    fi = _face_info;
  const bool use_elem = fi->faceType(_var.name()) == FaceInfo::VarFaceNeighbors::ELEM;
```

lindsayad reviewed (changes requested):
I don't see why this can't be ported up to `FVBoundaryCondition` and then we don't have two pretty much identical implementations in the derived classes

lindsayad reviewed (dismissed):
I really like the new changes. I'm going to do the right thing and approve this even though I'm giving myself more work again 😄

lindsayad commented on modules/navier_stokes/src/fvbcs/INSFVSymmetryPressureBC.C:
```diff
@@ -11,14 +11,15 @@
-  params.addClassDescription("Though applied to the pressure, this object ensures that the "
-                             "velocity perpendicular to a symmetry bounadry is zero by setting the "
-                             "mass flow rate across the symmetry boundary to zero.");
+  params.addClassDescription("Though not applied to velocity, this object ensures that the "
+                             "velocity perpendicular to a symmetry boundary is zero by setting "
+                             "the mass/scalar flow rate across the symmetry boundary to zero.");
```
I'm still not sure I love the description. Setting the scalar flux to zero at the boundary just requires that the product of rho, velocity, and the scalar concentration is zero. This could be accomplished by either the velocity *or* the scalar concentration being zero, so this object applied to a scalar flow doesn't guarantee that the velocity will be zero. Only application of this object to the pressure will ensure zero velocity perpendicular to the symmetry boundary since in the product of rho and velocity, only velocity can be zero.

GiudGiud replied:
Ok I ll fix that

GiudGiud commented:
haha ok. I ll make the exodus files lighter first.

GiudGiud commented:
Changed explanation and remove all but the final time steps on the exodus files. Should make them much lighter.

lindsayad commented:
@GiudGiud merge when you're ready
